### PR TITLE
`use-snapshots` build option for cross compile support.

### PR DESCRIPTION
**Pull request by afinch7**
For cross compiled builds it is [currently impossible to generate compatible snapshots at compile time](https://bugs.chromium.org/p/chromium/issues/detail?id=803591). The snapshots need to be generated on the platform they are intended to run on.

<strike>Currently it compiles fine, but fails generating the snapshot during runtime with the following error:
```
{"message":"Uncaught SyntaxError: Invalid regular expression flags","sourceLine":"/deno/target/release/gen/bundle/compiler.jsvar denoMain = (function () {","scriptResourceName":"/deno/target/release/gen/bundle/main.js","lineNumber":9468,"startPosition":415154,"endPosition":415155,"errorLevel":8,"startColumn":0,"endColumn":1,"isSharedCrossOrigin":false,"isOpaque":false,"frames":[]}
```

It also still needs to be designed in a way that works for all cross builds. I used annotations like this just to test things.
```
#[cfg(target_arch = "aarch64")]
```
</strike>

### CODE REVIEW

File: libdeno/api.cc [hunk 1]
```diff
@@ -108,10 +108,11 @@ deno_buf deno_get_snapshot(Deno* d_) {
 static std::unique_ptr<v8::Platform> platform;
 
 void deno_init() {
-  CHECK_NULL(platform.get());
-  platform = v8::platform::NewDefaultPlatform();
-  v8::V8::InitializePlatform(platform.get());
-  v8::V8::Initialize();
+  if (platform.get() == nullptr) {
```
Comment (afinch7) [inline libdeno/api.cc]: I'm not sure if this will cause problems. It seems to work fine this way, but most of the time there is a reason for things like this.

Comment (ry) [reply libdeno/api.cc]: I'm not sure either - but if the tests pass (which they do) then I'm ok with it.

File: src/snapshot.rs [hunk 2]
```diff
@@ -1,6 +1,37 @@
...
+  let cjs_file = CString::new(js_file).unwrap();
+  let cjs_source =
+    CString::new(std::str::from_utf8(js_source).unwrap()).unwrap();
+
+  v8_set_flags(vec![
+    "gen/snapshot_deno.bin".to_string(),
+    js_file.to_string(),
+  ]);
+
+  unsafe { deno_generate_snapshot(cjs_file.as_ptr(), cjs_source.as_ptr()) }
```
Comment (ry) [inline src/snapshot.rs]: This is generating the snapshot at runtime, which seems not so ideal. I think just executing the code instead would work..?

**Comment by ry** [comment]
This is great - I'd love to land this.

Am I reading this right that deno would generate a snapshot every time it starts up? This is should be changed to just startup without snapshots. We actually need this feature for warming up V8 before snapshotting to retain optimized code - so I would be happy to do the preliminary work.

It would be cool to get this building in CI so that we can retain compatibility and potentially ship a binary.

File: src/snapshot.rs (hunk 2 above)
Comment (afinch7) [reply src/snapshot.rs]: I want to save the snapshot to disk on first run, so it can be loaded from disk on subsequent runs.

Comment (ry) [reply src/snapshot.rs]: Hm... but then why not just include it in the executable?
We should try to get snapshot_creator to build as a target_arch...

Comment (ry) [reply src/snapshot.rs]: I guess we need something like this to accomplish that:
https://github.com/denoland/deno_third_party/blob/7e697531c16ee6713fa3d986447bbd4dc04a80bb/v8/BUILD.gn#L3492

Comment (afinch7) [reply src/snapshot.rs]: > Hm... but then why not just include it in the executable?
 We should try to get snapshot_creator to build as a target_arch...

How would one go about executing the snapshot generator on the compiler host if the snapshot generator is compiled for a different architecture.

**Comment by afinch7** [comment]
I still need to add some code to cache snapshots in the .deno directory so that after the first run we can use snapshots for each subsequent run on cross compiled builds.

**Comment by ry** [comment]
It would need to startup an emulator, I guess. I would need to research what V8 does with mksnapshot. Presumably they have a buildbot for cross compilation.

I'm not into the idea of caching the snapshot - this is inefficient and adds complexity.

I'm ok if we side step the snapshot completely for arm for now, until we can figure out how to generate the snapshot at compile time. We need this functionality in other places.

**Comment by afinch7** [comment]
That should work fine, but for some reason I the `use-snapshots` feature is always enabled. I commented it out for now so I could test on arm.

File: build.rs [hunk 3]
```diff
@@ -76,6 +76,10 @@ fn main() {
     }
   }
 
+  if cfg!(target_arch = "x86_64") {
```
Comment (afinch7) [inline build.rs]: For some reason this doesn't work like I expect. Instead of only enabling `use-snapshots` only for x86_64 builds it seems to be enabling them for all builds.

File: libdeno/api.cc [hunk 4]
```diff
@@ -225,4 +228,22 @@ void deno_terminate_execution(Deno* d_) {
...
+  Deno* d = deno_new(config);
+  deno_execute(d, nullptr, js_file, js_source);
+  if (deno_last_exception(d) != nullptr) {
+    std::cerr << "Snapshot Exception " << std::endl;
+    std::cerr << deno_last_exception(d) << std::endl;
+    deno_delete(d);
+    exit(1);  // TODO(afinch7): better error reporting here
+  }
+  auto snapshot = deno_get_snapshot(d);
+  deno_delete(d);
```
Comment (ry) [inline libdeno/api.cc]: This method is essentially a combination of `deno_init`, `deno_new`, `deno_execute`, `deno_last_exception`, `deno_delete`, and `deno_get_snapshot` - which prints the error to stderr and exits.

Since it does not introduce any new V8 bindings, I do not think this should be a top-level libdeno function.

Can you remove this from deno.h ?

**Comment by afinch7** [comment]
Made the requested changes. This should be ready to go.

File: build.rs [hunk 5]
```diff
@@ -76,6 +76,11 @@ fn main() {
     }
   }
 
+  // Enable snapshots for x64 builds
+  if env::var("CARGO_CFG_TARGET_ARCH").unwrap() == "x86_64" {
+    println!("cargo:rustc-cfg=feature=\"use-snapshots\"");
```
Comment (ry) [inline build.rs]: This `use-snapshots` cfg is a bit confusing to me because there's also a GN option called `v8_use_snapshot` which is not effected by this.
Could you add a comment here to this effect that these are unrelated? Maybe rename `use-snapshots` to `load-snapshots` ?

File: src/isolate.rs [hunk 6]
```diff
@@ -162,11 +163,21 @@ pub struct Metrics {
...
 
+pub struct IsolateInitScript {
+  pub source: String,
+  pub filename: String,
+}
+
+pub struct IsolateInit {
+  pub snapshot: Option<libdeno::deno_buf>,
+  pub init_script: Option<IsolateInitScript>,
+}
```
Comment (ry) [inline src/isolate.rs]: Move these structs to isolate_init.rs ?

File: src/isolate.rs [hunk 7]
```diff
@@ -176,11 +187,30 @@ impl Isolate {
...
+      Some(init_script) => {
+        let filename_cstr = CString::new(init_script.filename).unwrap();
+        let source_cstr = CString::new(init_script.source).unwrap();
+        unsafe {
+          libdeno::deno_execute(
+            libdeno_isolate,
+            null(),
+            filename_cstr.as_ptr(),
+            source_cstr.as_ptr(),
+          );
```
Comment (ry) [inline src/isolate.rs]: Why not use Isolate::execute2 instead of adding another call to `libdeno::deno_execute`? I'm concerned there might be (now or in the future) extra bits added to `Isolate::execute` which then would not be executed in this call?

Comment (afinch7) [reply src/isolate.rs]: Would doing this after creating the isolate struct be fine?

Comment (ry) [reply src/isolate.rs]: If that works for cross compiling and could simplify the IsolateInit arg - that would be preferable.

**Comment by afinch7** [comment]
Anything else?

**Comment by kevinkassimo** [comment]
Worth noting: this commit reduces Deno binary size, while doubling code startup time:
screen shot 2019-03-04 at 11 15 55 pm

**Comment by ry** [comment]
Oops. Not sure why this slow down is happening - but we must revert #1885

**Comment by afinch7** [comment]
I guess I'm going to need to do some investigation on this one.

**Comment by ry** [comment]
It seems likely that the snapshot isn’t being used on x64... but I haven’t dug in

**Comment by afinch7** [comment]
Ok first clue here.
`cargo build --release`
works fine `target/release/deno tests/002_hello.ts` ~13ms locally
`./tools/build.py -C target/release`
not fine same benchmark ~24ms

**Comment by afinch7** [comment]
You are in fact correct @ry, but for some reason this only happens when using build.py.

**Comment by afinch7** [comment]
@ry Found a fix going to create a pr now.
